Xpressive plugin

gnudles opened the pull request:
Xpressive plugin allows you to write your own expressions for sound creation.

AndiEcker commented:
+1 - WOW - looks like another interesting instrument

2017-01-14 19:53 GMT+00:00 gnudles <notifications@github.com>:

> Xpressive plugin allows you to write your own expressions for sound
> creation.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/LMMS/lmms/pull/3259
> Commit Summary
>
>    - Small fixes:
>    - First Preview of the X-Pressive Plugin
>    - Xpressive Plug-In:
>    - X-Pressive Plug-In: Few more fixes
>    - Special version of ExprTk with modified license (BSL) for the LMMS
>    project https://github.com/LMMS/lmms
>    - Merge pull request #1 from ArashPartow/xpressive_plugin
>
> File Changes
>
>    - *A* data/presets/X-Pressive/X-Pressive1.xpf
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-0> (21)
>    - *A* data/presets/X-Pressive/X-Pressive10.xpf
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-1> (21)
>    - *A* data/presets/X-Pressive/X-Pressive11.xpf
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-2> (21)
>    - *A* data/presets/X-Pressive/X-Pressive12.xpf
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-3> (21)
>    - *A* data/presets/X-Pressive/X-Pressive13.xpf
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-4> (21)
>    - *A* data/presets/X-Pressive/X-Pressive14.xpf
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-5> (21)
>    - *A* data/presets/X-Pressive/X-Pressive2.xpf
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-6> (21)
>    - *A* data/presets/X-Pressive/X-Pressive3.xpf
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-7> (21)
>    - *A* data/presets/X-Pressive/X-Pressive4.xpf
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-8> (21)
>    - *A* data/presets/X-Pressive/X-Pressive5.xpf
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-9> (21)
>    - *A* data/presets/X-Pressive/X-Pressive6.xpf
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-10> (21)
>    - *A* data/presets/X-Pressive/X-Pressive7.xpf
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-11> (21)
>    - *A* data/presets/X-Pressive/X-Pressive8.xpf
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-12> (21)
>    - *A* data/presets/X-Pressive/X-Pressive9.xpf
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-13> (21)
>    - *A* data/presets/X-Pressive/watsyn_like.xpf
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-14> (21)
>    - *M* include/Graph.h
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-15> (1)
>    - *M* plugins/CMakeLists.txt
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-16> (1)
>    - *A* plugins/xpressive/CMakeLists.txt
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-17> (4)
>    - *A* plugins/xpressive/artwork.png
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-18> (0)
>    - *A* plugins/xpressive/expressive_plugin.cpp
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-19> (885)
>    - *A* plugins/xpressive/expressive_plugin.h
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-20> (0)
>    - *M* plugins/xpressive/exprfront.cpp
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-21> (0)
>    - *M* plugins/xpressive/exprfront.h
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-22> (0)
>    - *M* plugins/xpressive/exprtk.hpp
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-23> (0)
>    - *M* plugins/xpressive/help_active.png
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-24> (0)
>    - *M* plugins/xpressive/help_inactive.png
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-25> (0)
>    - *M* plugins/xpressive/logo.png
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-26> (0)
>    - *M* plugins/xpressive/o1_active.png
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-27> (0)
>    - *M* plugins/xpressive/o1_inactive.png
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-28> (0)
>    - *M* plugins/xpressive/o2_active.png
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-29> (0)
>    - *M* plugins/xpressive/o2_inactive.png
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-30> (0)
>    - *M* plugins/xpressive/w1_active.png
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-31> (0)
>    - *M* plugins/xpressive/w1_inactive.png
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-32> (0)
>    - *M* plugins/xpressive/w2_active.png
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-33> (0)
>    - *M* plugins/xpressive/w2_inactive.png
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-34> (0)
>    - *M* plugins/xpressive/w3_active.png
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-35> (0)
>    - *M* plugins/xpressive/w3_inactive.png
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-36> (0)
>    - *M* plugins/xpressive/wavegraph.png
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-37> (0)
>    - *M* src/gui/PianoView.cpp
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-38> (0)
>    - *M* src/gui/TrackContainerView.cpp
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-39> (0)
>    - *M* src/gui/editors/AutomationEditor.cpp
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-40> (0)
>    - *M* src/gui/widgets/Graph.cpp
>    <https://github.com/LMMS/lmms/pull/3259/files#diff-41> (0)
>
> Patch Links:
>
>    - https://github.com/LMMS/lmms/pull/3259.patch
>    - https://github.com/LMMS/lmms/pull/3259.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/LMMS/lmms/pull/3259>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADlPbl76fEXaRN1LIu-LmwDd1v8B7wR9ks5rSSfPgaJpZM4LjuVX>
> .
>

zonkmachine commented:
![xpressive](https://cloud.githubusercontent.com/assets/6368949/22018292/68d51782-dcae-11e6-81f4-4c11d4f70721.png)

:open_mouth: Totally tripping right now. It's so cool. Like if Bit Invader grew up and took classes in engineering.

gnudles commented:
What should be done with this error? :
/tmp/cc62tQLz.s: Fatal error: can't write CMakeFiles/xpressive.dir/exprfront.cpp.obj: File too big

ArashPartow commented:
> What should be done with this error? :

The version of ld shipped with mingw64 is **Literally Broken** - last good version was circa GCC4.3.  In short to get rid of the build error you need to do the following two things:

**Step 1.** 
Add the following only to the mingw64 build options in CMake:

**`-Wa,-mbig-obj`**


**Step 2.** 
In order to make the ExprTk TU section size smaller add the following defines either in the mingw64 CMake build line 
 
- -Dexprtk_disable_break_continue
- -Dexprtk_disable_sc_andor
- -Dexprtk_disable_enhanced_features
- -Dexprtk_disable_string_capabilities
- -Dexprtk_disable_rtl_io_file

Or **#ifdef** on mingw64 just prior to the **#include "exprtk.hpp"** and define 

```
#define exprtk_disable_break_continue
#define exprtk_disable_sc_andor
#define exprtk_disable_enhanced_features
#define exprtk_disable_string_capabilities
#define exprtk_disable_rtl_io_file
```

gnudles commented:
oops, I'll replace
`IF(MINGW_PREFIX)`
with
`IF(WIN64)`
I invoke the parser with:
```
	parser_t::settings_store sstore;
	sstore.disable_all_logic_ops();
	sstore.disable_all_assignment_ops();
	sstore.disable_all_control_structures();
	parser_t parser(sstore);
```
So I don't understand why these defines should be used only in a certain build. Furthermore, I added them to reduce build time.

tresf commented:
@ArashPartow Thanks for such specific insight into this issue.  From what I'm reading, we can set **Step 1** globally, right?

https://github.com/LMMS/lmms/blob/master/cmake/modules/MinGWCrossCompile.cmake#L17-L22

, and

https://github.com/LMMS/lmms/blob/master/cmake/modules/MinGWCrossCompile.cmake#L61-L67

ArashPartow commented:
I think it should be something like:

`IF(MINGW_PREFIX and WIN64)`

As **WIN64** probably implies **ALL** windows 64 builds, perhaps including MSVC


-------------------


Perhaps online _assignment_ and _loop control_structures_ as you might want if-statements/ternary-statements with logical operators to be made available:

```
parser_t::settings_store sstore;
sstore.disable_all_assignment_ops();
sstore.disable_control_structure(settings_t::e_ctrl_for_loop   );
sstore.disable_control_structure(settings_t::e_ctrl_while_loop );
sstore.disable_control_structure(settings_t::e_ctrl_repeat_loop);
parser_t parser(sstore);
```


-------------------


> So I don't understand why these defines should be used only in a certain build. 

It's explained in my post **above**.


-------------------


> I do want all the users to be able to use
> (t>0)&(t<1)
> I saw that using the flag exprtk_disable_sc_andor changes the implementation of the "&" and "|" signs.

Yeah get rid of it from the build line. It wont save much for section size issues and it is probably better off to have it available/present.

ArashPartow commented:
@tresf  Nothing should be global. All of this is related to a broken version of  **mingw64** - So all the changes must only be specific to 64-bit builds using mingw64 and no other compilers or architectures. 

The build changes presented are a hack/kludge to reduce the section size of the TU so as to allow the broken **ld** that is shipped with **mingw64**  to build the final executable binary. 

The **_defines_** disable certain features such as expression optimizations so as to reduce the section size.

tresf commented:
> @tresf Nothing should be global. All of this is related to a broken version of mingw64 - So all the changes must only be specific to 64-bit builds using mingw64 and no other compilers or architectures.

Please look at the lines that are linked.  https://github.com/LMMS/lmms/pull/3259#issuecomment-273540427

ArashPartow commented:
ooh yeah I missed this line:

https://github.com/LMMS/lmms/blob/master/cmake/modules/MinGWCrossCompile.cmake#L10

Should be all good if it is only within that  **IF(LMMS_BUILD_MSYS)**  scope.

tresf commented:
> Should be all good if it is only within that IF(LMMS_BUILD_MSYS) scope.

Great but to be precise, I believe it needs to be for both scopes linked or moved into a separate conditional.

ArashPartow commented:
@tresf Now I'm confused. According to the build artifacts the 32-bit mingw build was successful however the 64-bit mingw build **failed**.

On line:

https://github.com/LMMS/lmms/blob/master/cmake/modules/MinGWCrossCompile.cmake#L18

The "IF(WIN64)" seems to suggest 64-bit but the comments around and in the scope and the "COMPILER32" stuff seem to suggest 32-bit. 

I'll let you guys figure it out. However you end up doing it my recommendation is that it should only apply to 64-bit builds using mingw64.

tresf commented:
Correct, we explicitly build the 32-bit host wrapper on 64-bit.  Naturally, the description would have to change if this were added.

softrabbit commented:
Actually, this shouldn't be (just) a plugin. @gnudles, could you get the expression evalution worked into the standard oscillator class(es)? That way it could be used in at least TripleOscillator or Organic as well, maybe some other plugin too.

(I haven't tried this plugin yet, but I like the idea.)

gnudles commented:
@softrabbit , I think that the next step is to create a sound scripting language, compile it with llvm (or some other jit engine) on runtime, and make a generic plug-in infrastructure... But then you need to add llvm to the dependency list...
@tresf what's the bottom line? 
@zonkmachine - it's better to use `integrate(f)` instead of `t*f` because it stays stable with frequency change (with base detuning). integrate(f) holds internal counter that counts `f` and return the sum divided by sample_rate.

zapashcanon commented on plugins/xpressive/exprfront.cpp:
```diff
@@ -0,0 +1,483 @@
+		}
+		int m_rseed;
+	};
+}*/
```
A comment to explain why this is commented would be better I guess.

zapashcanon commented on plugins/xpressive/exprfront.cpp:
```diff
@@ -0,0 +1,483 @@
+		const int data_size=sizeof(random_data)/sizeof(int);
+		int xi=(int)index;
+		int si=m_rseed%data_size;
+		int sa=m_rseed/data_size;
```
```c++
const int xi = [...];
const int si = [...];
const int sa = [...];
```

zapashcanon commented on plugins/xpressive/exprfront.cpp:
```diff
@@ -0,0 +1,483 @@
+		{
+			return vec[ix];
+		}
+		int ixnext=(ix+1)%size;
```
```c++
const int ixnext=(ix+1)%size;
```

zapashcanon commented on plugins/xpressive/exprfront.cpp:
```diff
@@ -0,0 +1,483 @@
+		//to disable interpolation uncomment:
+		//return static_cast<WaveSample*>(w)->samples[ix];
+
+		float xfrc=fraction(x);
```
```c++
const float xfrc=fraction(x);
```

zapashcanon commented on plugins/xpressive/exprfront.cpp:
```diff
@@ -0,0 +1,483 @@
+		T x=positiveFraction(index);
+		x=size*x;
+
+		int ix=(int)x;
```
```c++
const int ix=(int)x;
```

zapashcanon commented on plugins/xpressive/exprfront.cpp:
```diff
@@ -0,0 +1,483 @@
+	inline T operator()(const T& index)
+	{
+		T x=positiveFraction(index);
+		x=size*x;
```
```c++
const T x = size * positiveFraction(index);
```

zapashcanon commented on plugins/xpressive/expressive_plugin.cpp:
```diff
@@ -0,0 +1,884 @@
+	if (smoothness>0)
+	{
+		int guass_size=smoothness*5;
+		guass_size|=1;
```
```c++
const int guass_size = (smoothness * 5) | 1;
```

zapashcanon commented on plugins/xpressive/expressive_plugin.cpp:
```diff
@@ -0,0 +1,884 @@
+	{
+		int guass_size=smoothness*5;
+		guass_size|=1;
+		int guass_center=guass_size/2;
```
```c++
const int guass_center=guass_size/2;
```

zapashcanon commented on plugins/xpressive/expressive_plugin.cpp:
```diff
@@ -0,0 +1,884 @@
+		int guass_size=smoothness*5;
+		guass_size|=1;
+		int guass_center=guass_size/2;
+		float delta=smoothness;
```
```c++
const float delta=smoothness;
```

zapashcanon commented on plugins/xpressive/expressive_plugin.cpp:
```diff
@@ -0,0 +1,884 @@
+		guass_size|=1;
+		int guass_center=guass_size/2;
+		float delta=smoothness;
+		float a=1/(sqrtf(2*F_PI)*delta);
```
```c++
const float a=1/(sqrtf(2*F_PI)*delta);
```

zapashcanon commented on plugins/xpressive/expressive_plugin.cpp:
```diff
@@ -0,0 +1,884 @@
+
+		if (parse_ok) {
+			e->m_exprValid.setValue(0);
+			int length = m_raw_graph->length();
```
```c++
const int length = m_raw_graph->length();
```

zapashcanon commented on plugins/xpressive/expressive_plugin.cpp:
```diff
@@ -0,0 +1,884 @@
+		}
+		expr.setIntegrate(&i,sample_rate);
+
+		bool parse_ok=expr.compile();
```
```c++
const bool parse_ok=expr.compile();
```

zapashcanon commented on plugins/xpressive/expressive_plugin.cpp:
```diff
@@ -0,0 +1,884 @@
+		ExprFront expr(text.constData());
+		float t,f=10,key=2,rel=0,v=0.5;
+		unsigned int i;
+		unsigned int sample_rate=m_raw_graph->length();
```
```c++
float t;
const float f=10,key=2,rel=0,v=0.5;
unsigned int i;
const unsigned int sample_rate=m_raw_graph->length();
```

zapashcanon commented on plugins/xpressive/expressive_plugin.cpp:
```diff
@@ -0,0 +1,884 @@
+		float t,f=10,key=2,rel=0,v=0.5;
+		unsigned int i;
+		unsigned int sample_rate=m_raw_graph->length();
+		expr.add_variable("t", t);
```
`t` not initialized ?

gnudles replied:
It works, but I'll initialize it for clarity.

gnudles replied:
I mean, the value is set before it's evaluated...

zapashcanon commented on plugins/xpressive/expressive_plugin.cpp:
```diff
@@ -0,0 +1,884 @@
+
+static void clearGraph(graphModel * g)
+{
+	int length = g->length();
```
```c++
const int length = g->length();
```

zapashcanon commented on plugins/xpressive/expressive_plugin.cpp:
```diff
@@ -0,0 +1,884 @@
+	m_panning2.loadSettings(_this,"PAN2");
+	m_relTransition.loadSettings(_this,"RELTRANS");
+
+	int size = 0;
```
Not sure for this one, as I don't know the effect of the next function calls, but:

```c++
const int size = 0;
```

gnudles replied:
Let's keep it that way...

zapashcanon commented on plugins/xpressive/expressive_plugin.cpp:
```diff
@@ -0,0 +1,884 @@
+
+	float o1,o2,pn1,pn2;
+	float new_freq=nph->frequency();
+	float freq_inc=(new_freq-frequency)/_frames;
```
```c++
const float new_freq=nph->frequency();
const float freq_inc=(new_freq-frequency)/_frames;
```

tresf commented:
Is this ready for merge?

zonkmachine commented:
> Is this ready for merge?

I think it is.

I've seen a very minor issue. The presets are not saved with unique names. It's not obvious when you save a preset that the file name and the preset name isn't necessarily the same. One could argue that this is something we could change actually.

zonkmachine commented on data/presets/X-Pressive/X-Pressive1.xpf:
```diff
@@ -0,0 +1,21 @@
+<!DOCTYPE lmms-project>
+<lmms-project version="1.0" creator="LMMS" creatorversion="1.2.0-rc2.6" type="instrumenttracksettings">
+  <head/>
+  <instrumenttracksettings muted="0" type="0" name="X-Pressive" solo="0">
```
`name="X-Pressive"`

The preset name that shows up when you load an instrument isn't the name of the `.xpf` but this setting. You need to edit the name in the rename dialogue on the top of the plugin.

gnudles replied:
Tell you the truth, I'm not sure that those presets should be merge. And even if they do, I think they should have more meaningful naming... I'm not so good at naming sounds so... if someone volunteers... 
Also if someone made nice presets which he wants them to be merged...

zonkmachine replied:
The user will want to have some good presets to look at for this one. In any case I don't think this should hold back a merge. You or someone else can tweak the names later and it's easier for people to try it out and make presets if it's merged anyway.

jasp00 commented on plugins/xpressive/CMakeLists.txt:
```diff
@@ -0,0 +1,7 @@
+INCLUDE(BuildPlugin)
+
+SET(CMAKE_CXX_FLAGS "-Dexprtk_disable_sc_andor -Dexprtk_disable_enhanced_features -Dexprtk_disable_break_continue -Dexprtk_disable_comments -Dexprtk_disable_string_capabilities -Dexprtk_disable_rtl_io_file -Dexprtk_disable_rtl_vecops ${WERROR_FLAGS}")
```
Please include previous value of `CMAKE_CXX_FLAGS`. You should apply all these definitions only for the MinGW-w64 build, as recommended at https://github.com/LMMS/lmms/pull/3259#issuecomment-273447592.

gnudles replied:
But I want to disable those features on any platform...

gnudles replied:
Oh... now it's not compiling because exprtk.hpp still throws exception in line 2733
@ArashPartow, should I just make a little fix like:
if (stride > 4)
            {
-               throw std::invalid_argument("token_scanner() - Invalid stride value");
+               stride = 4;
            }
????

ArashPartow replied:
@gnudles  which plaform are you building where you see that exception being thrown?

gnudles replied:
The exception is not being thrown. But I can't compile it with the flag -fno-exceptions

jasp00 commented on plugins/xpressive/expressive_plugin.cpp:
```diff
@@ -0,0 +1,926 @@
+/*
+ * expressive_plugin.cpp - instrument which uses a mathematical formula parser
+ *
+ * Copyright (c) 2016-2017 Orr Dvori
```
The convention is to add an email address, but it is not strictly necessary.

jasp00 commented on plugins/xpressive/expressive_plugin.cpp:
```diff
@@ -0,0 +1,926 @@
+ *
+ * Copyright (c) 2016-2017 Orr Dvori
+ * 
+ * This file is part of LMMS - http://lmms.io
```
https.

gnudles replied:
The repository is full with this mistaken address...
It can be done once with find/replace.

zonkmachine replied:
It's was fixed a week ago so this is the only instance left with http:

jasp00 commented on plugins/xpressive/expressive_plugin.cpp:
```diff
@@ -0,0 +1,926 @@
+#include "SubWindow.h"
+#include "GuiApplication.h"
+#include "MainWindow.h"
+#include "exprfront.h"
```
Please sort headers if possible.

jasp00 commented on plugins/xpressive/expressive_plugin.cpp:
```diff
@@ -0,0 +1,926 @@
+
+exprSynth::exprSynth(const WaveSample *gW1, const WaveSample *gW2, const WaveSample *gW3, ExprFront *_exprO1, ExprFront *_exprO2,
+					NotePlayHandle *_nph, const sample_rate_t _sample_rate, const FloatModel* _pan1, const FloatModel* _pan2, float _rel_trans)
+	:exprO1(_exprO1),exprO2(_exprO2),W1(gW1),W2(gW2),W3(gW3),nph(_nph),sample_rate(_sample_rate),pan1(_pan1),pan2(_pan2),rel_transition(_rel_trans)
```
All kind of types begin with an upper case letter. Member variables are prefixed with `m_`.

Although this is work in progress, please follow the format style (one line per member) used in other parts of LMMS. Other developers may need to fix your code or help you.

jasp00 commented on plugins/xpressive/expressive_plugin.cpp:
```diff
@@ -0,0 +1,926 @@
+					NotePlayHandle *_nph, const sample_rate_t _sample_rate, const FloatModel* _pan1, const FloatModel* _pan2, float _rel_trans)
+	:exprO1(_exprO1),exprO2(_exprO2),W1(gW1),W2(gW2),W3(gW3),nph(_nph),sample_rate(_sample_rate),pan1(_pan1),pan2(_pan2),rel_transition(_rel_trans)
+{
+	note_sample=0;
```
Infix operators should have a space before and after.

jasp00 commented on plugins/xpressive/expressive_plugin.cpp:
```diff
@@ -0,0 +1,926 @@
+exprSynth::~exprSynth()
+{
+	if (exprO1)
+		delete exprO1;
```
You should use explicit blocking with `{` `}`.

jasp00 commented on plugins/xpressive/expressive_plugin.cpp:
```diff
@@ -0,0 +1,926 @@
+	if (exprO2)
+		delete exprO2;
+}
+float saw_wave(float);
```
It is a good idea to declare internal functions as `static`.

jasp00 commented on plugins/xpressive/expressive_plugin.cpp:
```diff
@@ -0,0 +1,926 @@
+#define GRAPH_LENGTH 4096
+
+expressive::expressive(InstrumentTrack * _instrument_track) :
+		Instrument(_instrument_track, &xpressive_plugin_descriptor), m_graphO1(
```
Please do not add excessive indentation.

gnudles replied:
What do you mean? those two tabs or the list below?

zonkmachine replied:
The tabs!

jasp00 commented on plugins/xpressive/expressive_plugin.h:
```diff
@@ -0,0 +1,254 @@
+
+	virtual f_cnt_t desiredReleaseFrames() const
+	{
+		return( 64 );
```
Return without parenthesis.

jasp00 commented on plugins/xpressive/expressive_plugin.h:
```diff
@@ -0,0 +1,254 @@
+	
+	BoolModel m_exprValid;
+	
+	friend class expressiveView;
```
Be aware this is not a clean model/view separation.

gnudles replied:
I saw it in bitInvader (I started by copying bitInvader)...

jasp00 replied:
If you have some spare time, please use inline getters and setters instead of `friend`.

jasp00 commented on plugins/xpressive/expressive_plugin.h:
```diff
@@ -0,0 +1,254 @@
+	static QString HelpText;
+
+};
+class SubWindow;
```
Please move these declarations to the top.

jasp00 commented on plugins/xpressive/expressive_plugin.cpp:
```diff
@@ -0,0 +1,926 @@
+		delete exprO2;
+}
+float saw_wave(float);
+float sin_wave(float);
```
Where are these functions used?

gnudles replied:
leftovers...

jasp00 commented on plugins/xpressive/exprfront.cpp:
```diff
@@ -0,0 +1,533 @@
+{
+	static inline float process()
+	{
+		return 1.0f - rand() * 2.0f / RAND_MAX;
```
`rand` is not thread-safe. It should be replaced later.

jasp00 commented on plugins/xpressive/exprfront.cpp:
```diff
@@ -0,0 +1,533 @@
+	size_t last_pos=0;
+	size_t count = 0;
+	size_t len=strlen(needle);
+	while (1)
```
You should use the standard true constant.

gnudles commented:
I fixed everything that @jasp00 noted,

> Please include previous value of CMAKE_CXX_FLAGS
 but unfortunately when I added the previous value, it couldn't be compiled, because there is still another throw statement in exprTk that slipped through all those disabling defines.

simonvanderveldt commented:
Cool stuff!
Small question: Shouldn't new plugins use the new styling? See #2831
Or will this be done after this PR is merged?

zonkmachine commented:
> I fixed everything that @jasp00 noted,

@gnudles Awesome! Push the changes.


> > Please include previous value of CMAKE_CXX_FLAGS
> 
> but unfortunately when I added the previous value, it couldn't be compiled, because there is still another throw statement in exprTk that slipped through all those disabling defines.
> When it will be fixed, I think we are ready for merge...


Is there an upstream issue about this we can keep track of?


> Small question: Shouldn't new plugins use the new styling? See #2831
> Or will this be done after this PR is merged?

#2831 Is concerning the native effect plugins specifically.

simonvanderveldt commented:
> #2831 Is concerning the native effect plugins specifically.

@zonkmachine yeah, I always forget there's a distinction between instruments and effects :x

gnudles commented:
> @gnudles Awesome! Push the changes.

But if I do, it won't compile

> Is there an upstream issue about this we can keep track of?

What do you mean?

jasp00 commented on plugins/xpressive/exprfront.cpp:
```diff
@@ -0,0 +1,610 @@
+0xa66f8b47
+};
+
+inline unsigned int rotate(unsigned int x, const int b)
```
You should call this `rotateLeft`. Also, you should check if `b == 0` because you could cause undefined behavior.

jasp00 commented:
@gnudles
Regarding `exprtk.patched.hpp`, if you use `assert()` and define `NDEBUG`, you silence errors. You should print a message to standard error, even once.

@zonkmachine
>I probably have the weakest hardware of the developers

Are you able to compile `exprfront.cpp`? It is too big for my system. I will not be able to help if we do not split the file.

gnudles commented:
@jasp00, it shouldn't reach there anyway...

jasp00 commented:
>it shouldn't reach there anyway...

So it will be a cold function. Could you write this with your words?
```
#define xAssert() perror("exprtk exception")
```

zonkmachine commented:
> Are you able to compile exprfront.cpp?

I can compile the latest commit and it runs fine, if that is what you meant.

jasp00 commented:
>I can compile the latest commit and it runs fine

If you `touch exprfront.cpp` and `time make`, how long does it take?

zonkmachine commented:
```
real	0m52.185s
user	0m48.320s
sys	0m1.800s
```

gnudles commented:
Maybe it should be an external plug-in?

jasp00 commented:
>Maybe it should be an external plug-in?

No, I will split the file if necessary.

gnudles commented:
I don't like to use exprtk.patched.hpp, I prefer to use the original, but currently I cannot do so. I'm waiting for @tresf or @ArashPartow for further guidelines.

tresf commented:
> Small question: Shouldn't new plugins use the new styling? See #2831

For effects plugins, certainly.  For instrument plugins like this one... Ideally, but not mandatory.

Why not mandatory?  Well, this would generally require the OP to grant commit access to his/her repository to a designer like @RebeccaDeField requiring coordination/collaboration.  I'll defer that question to the OP and @RebeccaDeField if that can happen prior to merge.  If not, we can follow up with this task afterward.

> I don't like to use exprtk.patched.hpp, I prefer to use the original, but currently I cannot do so. I'm waiting for @tresf or @ArashPartow for further guidelines.

Can you or @jasp00 please summarize what's left (or the problem) on this file/task?  If the patch can be automated, @jasp00 has written some nice upgrade routines in other PRs we can leverage but if it's too complex to do, I'd like to know the harm in bundling the patched file.

We can also ship a `.diff` file with the `.hpp` and use that at compile time as often the `.diff` is easier to merge with upstream changes over a patched snapshot.

But I'm just speaking hypothetically as I don't know enough about the history of this source file to be able to help.  Perhaps my name was tagged accidentally?

gnudles commented:
@tresf, the question is - do I have to respect the -fnoexceptions flag, because it forces me to make small changes in the exprtk utility.

tresf commented:
> 

@jasp00 does some custom cmake tasks on his unmerged `clang` branch.  The method would bear some similarities.

First, he issues a custom cmake [`COMMAND`](https://github.com/jasp00/lmms/blob/clang/plugins/GigPlayer/CMakeLists.txt#L15-L34) which could be adapted to apply a diff.

Second, he shims the custom file into the [`BUILD_PLUGIN`](https://github.com/jasp00/lmms/blob/clang/plugins/GigPlayer/CMakeLists.txt#L38-L42) command.

His is configurable due to the reliance on a system header.  Yours can be hard-coded.

tresf commented on plugins/xpressive/CMakeLists.txt:
```diff
@@ -1,7 +1,16 @@
-BUILD_PLUGIN(xpressive expressive_plugin.cpp exprfront.cpp expressive_plugin.h MOCFILES expressive_plugin.h EMBEDDED_RESOURCES "${CMAKE_CURRENT_SOURCE_DIR}/*.png")
+#file(GENERATE OUTPUT "${CMAKE_CURRENT_BINARY_DIR}/exprtk.hpp"
+#	INPUT "${CMAKE_CURRENT_SOURCE_DIR}/exprtk.hpp.origin")
+file(COPY "${CMAKE_CURRENT_SOURCE_DIR}/exprtk.hpp.origin" DESTINATION "${CMAKE_CURRENT_BINARY_DIR}")
```
`exprtk.hpp` is just fine.  No need to call it `origin`.

gnudles replied:
I renamed it in order that it won't be included by the preprocessor...

tresf replied:
@gnudles I'm pretty sure the `cmake` configure step runs before all of that.  Can you explain?

gnudles replied:
I'm sorry, I don't understand the question...
If you have better place for the cmake patch script then tell me...

tresf commented on plugins/xpressive/exprtk_patch:
```diff
@@ -0,0 +1,64 @@
+57,58c57,69
```
`exprtk.hpp.diff` or `exprtk.diff` would be the accepted file extension.  This will trigger syntax highlighting in most code editors.

tresf commented:
@gnudles this looks great.  I left a few comments.

I've been thinking about the best place to put the patched `.hpp` file and I'm interested to know opinions on this:

* **ORIGINAL**: `plugins/xpressive/exprtk.hpp`
* **PATCH**: `plugins/xpressive/exprtk.hpp.diff`
* **COMPILE**: `plugins/xpressive/exprtk.patched.hpp`
* **GITIGNORE**:  `plugins/xpressive/.gitignore` entry: `*.patched.*`

Apologies for the slight change of direction.  It think leaving it in the source directory but omitting it from git is cleanest.

RebeccaDeField commented:
> I'll defer that question to the OP and @RebeccaDeField if that can happen prior to merge. If not, we can follow up with this task afterward.

I am currently busy wrapping up the effects plugins with some other contributors. I would suggest that if we want to redesign all of the other plugins, we follow up with that project later and once [this](#2831) is done. We would want to decide on some consistent design conventions for those plugins (I will go into this later because I want to stay on topic), I think this can be pulled in without a redesign for the time being.

jasp00 commented:
>I've been thinking about the best place to put the patched `.hpp` file

Please be simple, just use one consolidated file. Prior to merging, we should keep the original file around to make a patch for upstream.

>do I have to respect the -fnoexceptions flag

@gnudles, try and see if you can use exceptions.

gnudles commented:
I don't know why it stopped compile. Any suggestions? it looks like the PLUGIN_NAME define is being ignored...

jasp00 commented:
>Any suggestions?

Please edit `cmake/modules/BuildPlugin.cmake` to stop including `${CMAKE_SOURCE_DIR}/src/gui`. Then include `embed.h` instead of `embed.cpp`.

Umcaruje commented:
So we don't want this plugin in 1.2 after all?

jasp00 commented:
From [build log](https://travis-ci.org/LMMS/lmms/jobs/218398018),
```
Fatal error: can't write CMakeFiles/xpressive.dir/exprsynth.cpp.obj: File too big
```
Please chop the code.

gnudles commented:
yeah, it should be ok now... I tried to enable the variable-constant optimizations (exprtk_disable_enhanced_features).
I also moved the patching process to occur in the build time.

gnudles commented:
I think it's ready to be merged... Isn't it?

jasp00 commented:
>I also moved the patching process to occur in the build time.

[Please](https://github.com/LMMS/lmms/pull/3259#issuecomment-289730206) be simple, use consolidated files while keeping the original header.

What error do you get if you remove `-Dexprtk_disable_` flags from the normal build?

gnudles commented:
> What error do you get if you remove -Dexprtk_disable_ flags from the normal build?

It make the compilation process slower and it adds more throw statements...

> Please be simple, use consolidated files while keeping the original header.
But @tresf said that it's OK...

gnudles commented:
BTW, the error in the build system is not my fault...

jasp00 commented:
>It make the compilation process slower

But it yields more features, right? Compilation speed may be decreased by splitting logic.

>and it adds more throw statements...

Which you know how to handle. Did you try to build with exceptions? Does ExprTk use any callbacks?

>But @tresf said that it's OK...

As @tresf [points out](https://github.com/LMMS/lmms/pull/3259#issuecomment-289231753), we have full control over the bundled source. We are not using `dpatch` or `quilt`, thus working with patches is more complicated. So please be simple and use consolidated files.

gnudles commented:
> Does ExprTk use any callbacks?

Yes. it receives some callbacks but I only use callbacks in polymorphism manner, I mean through a class overridden method, and not as a bare function. 
Remeber when you told me on the walk-through :
> Please include previous value of CMAKE_CXX_FLAGS

Beforehand, when it was without the previous value of CMAKE_CXX_FLAGS, it was compiled without the -fnoexceptions flag, and everything worked correctly, at least on my machine (linux 64bit). Maybe it poses some problems on Windows(R) or OS/X, although it shouldn't.
If exceptions will be turned on, I won't have to patch the lib ( except for the exprtk_disable_return flag that I added, but we can just ignore the fact that return statement exists... )

> But it yields more features, right?

I've already told you that those features are an overkill. I do not want to enable assignment operation, and enabling while-loops may lead to infinite loop and even if it won't, it's hard to render 44100*2 samples per second when you do too much operation without JIT compiler...
About the flags:
exprtk_disable_sc_andor - I'm not so sure about it because and/or operations still working without it... (maybe it does some optimizations).
exprtk_disable_return - I made that flag to disable return statement.
exprtk_disable_break_continue - disables some control statements
Dexprtk_disable_comments - disable parsing comments in the expression.
Dexprtk_disable_string_capabilities - disable parsing strings and manipulate them in the expression.
Dexprtk_disable_rtl_io_file - disable file I/O
Dexprtk_disable_rtl_vecops - disable vector operations... I haven't tried that feature, but I don't think it's useful in our case...

> Compilation speed may be decreased by splitting logic.

It's not so simple to split this file. You need to master all the inner dependencies and you should also remember that it's a templates file.
I know it's annoying to compile this file, and it's the largest source file in LMMS for sure, but you have the choise if to insert it or not.

tresf commented:
> Please be simple, just use one consolidated file. Prior to merging, we should keep the original file around to make a patch for upstream.

Patching is fine.  It's a clean 61-line diff.  I do agree we should add a `FIXME` or `TODO` as a reminder to file the patch upstream.

@jasp00 if you require this patched prior to merging, perhaps you can help direct us [where to submit it](http://www.partow.net/programming/exprtk/index.html).

The main reasoning for recommending to patch this file is @gnudles [didn't want to include the patched version](https://github.com/LMMS/lmms/pull/3259#issuecomment-288686998) and I don't blame him since the diff is small and the original source is huge.

That conversation happened 5 days ago, so I think we had assumed no news was good news. :)

gnudles commented:
The patch is combined from two main changes:
1) a workaround to compile with -fnoexception flag...
2) a flag to disable return statement... 
So AFAIK if you want to push the changes to upstream, only change #2 should be considered...
BUT! I talked with Arash, and he said he will do it ([change #2](https://github.com/gnudles/lmms/pull/2#issuecomment-291642100)) .

> perhaps you can help direct us where to submit it.

Perhaps via pull request in github?

tresf commented:
> Perhaps via pull request in github?

My apologies, I didn't realize @ArashPartow was the upstream author of the library.

Are you saying upstream can take all of our changes, or no?  If they can't, we should keep the diff.  If they can, we should PR. 👍

gnudles commented:
They shouldn't, because regular users of exprtk.hpp can use exceptions in their environment... ;)
And the other change is (probably) already being done by the author.
Just to be sure, when you say "upstream" you mean the exprtk repository, managed by Arash, right?

tresf commented on plugins/xpressive/CMakeLists.txt:
```diff
@@ -0,0 +1,14 @@
+INCLUDE(BuildPlugin)
+
+SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Dexprtk_disable_sc_andor -Dexprtk_disable_return -Dexprtk_disable_break_continue -Dexprtk_disable_comments -Dexprtk_disable_string_capabilities -Dexprtk_disable_rtl_io_file -Dexprtk_disable_rtl_vecops ${WERROR_FLAGS}")
+IF(WIN32 OR WIN64)
```
[`WIN32`](https://cmake.org/cmake/help/v3.0/variable/WIN32.html) actually covers both cases, `WIN64` can safely be removed.

tresf commented:
Ok, so I'm reading through the history of this and I feel @jasp00 can be more specific with the review process.

@jasp00 keeps stating things like "split the file", "keep it simple" yet @gnudles has clearly stated that not all of these patches aren't upstream worthy.

Can you please nudge this PR in the right direction?  If not, we'll merge as-is.  This PR is very clean and the `diff` file is sane and small.

If we can leverage exceptions for this plugin, please instruct @gnudles to do so.  If we can't, the `diff` is an acceptable approach because it will make for quicker merging from upstream.

At one point @softrabbit asked if we could make this library part of the base code.  Should that be handled in a separate PR?

> Just to be sure, when you say "upstream" you mean the exprtk repository, managed by Arash, right?

Yes.  And on a side note, thanks for the work on this so far.  @jasp00 has recently re-joined our team and brings a lot of expertise, so PRs are getting more attention to fine detail.  I hope that's ok. :)

ArashPartow commented:
@gnudles @tresf

> a flag to disable return statement...
> So AFAIK if you want to push the changes to upstream, only change # 2 should be considered...
> BUT! I talked with Arash, and he said he will do it (change #2) .


I think I may have misunderstood the _actual_ requirement here.

My understanding was you wanted to disable the use of the return statement in expressions. So i went and added an extra settings mode to the parser settings.

But reading this thread, it seems your intention was not only to disable return statements but to also allow for a build with **fno-exceptions**?

If that is correct, we'll probably need to add the if-defs @gnudles proposed earlier on. Please let me know if that's what you want to do.

tresf commented:
> If that is correct, we'll probably need to add the if-defs @gnudles proposed earlier on. Please let me know if that's what you want to do.

@ArashPartow thanks kindly for keeping track of this PR. :)  I had to re-read this as well... yes, it seems to be two-fold however I'll let @gnudles and @jasp00 make that call.

ArashPartow commented:
@tresf @gnudles 

I've added both [options](https://github.com/gnudles/lmms/pull/4#issue-219736929), you can choose to use whichever is most appropriate.

ArashPartow commented:
@gnudles

In the xpressive's CMakeLists.txt you have the following:

```
IF(WIN32 OR WIN64)
   SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wa,-mbig-obj -Dexprtk_disable_enhanced_features")
ENDIF()
```

My understanding was that defining exprtk_disable_enhanced_features was only required for the 32-bit MinGW version, and was not required for the 64-bit version.

Is this not the case anymore?

tresf commented:
> My understanding was that defining exprtk_disable_enhanced_features was only required for the 32-bit MinGW version, and was not required for the 64-bit version.

In that case perhaps... 

```diff
- IF(WIN32 OR WIN64)
+ IF(WIN32 AND NOT WIN64)
     SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wa,-mbig-obj -Dexprtk_disable_enhanced_features")
  ENDIF()
```

tresf commented:
Or perhaps... 

```cmake
IF(WIN32)
   SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wa,-mbig-obj")
   IF (NOT WIN64)
      SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Dexprtk_disable_enhanced_features")
   ENDIF()
ENDIF()

ArashPartow commented:
@tresf the last option seems most appropriate.

gnudles commented:
> My understanding was that defining exprtk_disable_enhanced_features was only required for the 32-bit MinGW version, and was not required for the 64-bit version.

At the following build:
https://travis-ci.org/LMMS/lmms/builds/192966413
It failed on "file too large" on WIN64 so I disabled many features including the exprtk_disable_enhanced_features on ALL platforms (mistakenly).
But, recently I discovered that this flag only adds optimization... so I enabled the enhanced features on ALL platforms except WIN64. and then the WIN32 build failed with "file too large" error, so I disabled the enhanced features on both of them.

> Edit: or perhaps...

What @tresf suggests seems good.

> I've added both options, you can choose to use whichever is most appropriate.

Thanks a lot! (it indeed able to throw return_exception() if not disabled...)

CONCLUSIONS - 
I'll fix the CMAKE for the WIN64/32 flags.
I'll merge the new exprtk library.
@jasp00 should instruct me about the -fno-exceptions if he wants to...

gnudles commented:
> IF (NOT WIN64)

Wait, now I'm confused... @ArashPartow you said before that the ld of Mingw for WIN64 is broken, not for WIN32...

gnudles commented:
You can see in the log : https://travis-ci.org/LMMS/lmms/jobs/219417335
that the fail occurs on WIN64. I'm updating the Cmake as I suggested...

gnudles commented:
I must leave tomorrow's morning, and I won't have an internet connection for a while. Can somebody take control?

tresf commented:
@gnudles yeah, paraphrasing https://github.com/LMMS/lmms/pull/3259#issuecomment-273447592 that is what it would appear.

Note, to be as technical as possible... if this problem is with `mingw-w64`, the yes it would affect both `mingw-w64` 32-bit and `mingw-w64` 64-bit alike.  (Yes, it's [exactly as confusing](https://sourceforge.net/projects/mingw-w64/) as it reads).

And to nitpick, @ArashPartow has told us that this bug doesn't necessarily effect MSVC++, so to be proper, we should validate `MINGW_PREFIX`, as we may offer MSVC++ support in the next year or so.

tresf commented:
> Can somebody take control?

This rarely works with forked repositories, so if this needs to happen we would have to supercede your PR, or you'd have to grant us commit access on your fork.  If we know how long you'll be gone for, that would help the decision.

gnudles commented:
I hope it did it.
I might have a random access in about two weeks from now.

Umcaruje commented:
Github already allows this without you providing us access to your whole fork: 

https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/
https://help.github.com/articles/committing-changes-to-a-pull-request-branch-created-from-a-fork/

jasp00 commented:
>@jasp00 should instruct me about the -fno-exceptions if he wants to...

* https://github.com/LMMS/lmms/pull/3259#issuecomment-289730206: "@gnudles, try and see if you can use exceptions."
* https://github.com/LMMS/lmms/pull/3259#issuecomment-291482763: "Did you try to build with exceptions?"

>If we can leverage exceptions for this plugin, please instruct @gnudles how best to do so.

Build with `-fexceptions`. Do you know what a try-block is?

>I've already told you that those features are an overkill.

You do not like those features, fine, do not use them, but allow other users to choose.

>It's not so simple to split this file.

It is not that hard, take your time and ask for guidance, upstream is cooperating. You should be able to enable all features and maybe drop MinGW flags.

>diff is an acceptable approach

It is not the preferred form for making modifications, you do not edit the diff file. A simple alternative would be to ship the patched file and the diff to the original. Anyway, we can try not to have any diff.

tresf commented:
> Do you know what a try-block is?

No, not in C++ and not in LMMS.  I was hoping you could help out.

> It is not the preferred form for making modifications

For some projects it is.  MacPorts uses this technique and it's for identical purposes (to fix compilation or compatibility problems not yet acknowledged, fixed or intended to be included by upstream)

> Anyway, we can try not to have any diff.

I agree that sounds like the best option, but I will not prevent this feature from being merged on that reason alone.

jasp00 commented:
Are you taking care of this, @tresf?

A [try-block](http://en.cppreference.com/w/cpp/language/try_catch) is a sequence of one try-clause and one or more catch-clauses. For this plug-in, this should work:

```c++
	try
	{
		// ExprTk calls
	}
	catch(...)
	{
		qWarning("ExprTk exception");
	}
```

tresf commented:
Not yet.  I won't​ be able to for some weeks. If you have the bandwidth and desire, please do.

JordanKyser commented:
I see that this is an advanced BitInvader plug-in, and you can put your formula for a particular wave (as in soundwave, not the .wav file) sample. I like BitInvader, and I will sure use Xpressive when it is implemented in LMMS.

gnudles commented:
@ArashPartow, Compiling with some disable flags seem to cause some troubles...

ArashPartow commented:
Looks like a whole bunch of things were missed in the merge.  In any case I've made a new PR: [ 	e330a93](https://github.com/gnudles/lmms/pull/6)

zonkmachine commented:
It passes Travis, builds fine and works like a charm. :whale:

tresf commented on plugins/xpressive/.gitignore:
```diff
@@ -0,0 +1,2 @@
+#ignore the patched file
+exprtk.patched.hpp
```
Has the patched file been removed? (can this file be removed now?)

gnudles replied:
The patch was removed due to compilation with the -fexceptions flag and the release of new exprtk version that allow one to disable return statement.

gnudles replied:
All calls for Exprtk are now surrounded by try-blocks.

tresf commented on cmake/modules/BuildPlugin.cmake:
```diff
@@ -9,7 +9,7 @@ MACRO(BUILD_PLUGIN PLUGIN_NAME)
 	SET(PLUGIN_SOURCES ${PLUGIN_UNPARSED_ARGUMENTS})
 
-	INCLUDE_DIRECTORIES(${CMAKE_CURRENT_BINARY_DIR} ${CMAKE_BINARY_DIR} ${CMAKE_SOURCE_DIR}/include ${CMAKE_SOURCE_DIR}/src/gui)
+	INCLUDE_DIRECTORIES(${CMAKE_CURRENT_BINARY_DIR} ${CMAKE_BINARY_DIR} ${CMAKE_SOURCE_DIR}/include)
```
I'm sorry if this was already explained previously, but was `${CMAKE_SOURCE_DIR}/src/gui` deemed as unnecessary for all of our plugins and was lingering from something else?

tresf replied:
Update, this appears to [date back 10 years](https://github.com/gnudles/lmms/blob/e637495/cmake/modules/BuildPlugin.cmake#L50) and I see no evidence it's needed as headers aren't in that directory.  Should be safe to remove. 👍

gnudles commented:
I think it is ready to be merge. The only disadvantage is that you can't draw a wave shape by hand. (I disabled that possibility when I began working on this plug-in to make things simpler). Enabling that require a little work, and I don't have the time for that right now...
Anyway, I recently added the tempo variable to the evaluator, and this is awesome because you can make a tempo aware synthesizer.

Umcaruje commented:
@gnudles can you please change the base branch to `stable-1.2`, not master

zonkmachine commented:
> @gnudles can you please change the base branch to stable-1.2, not master

Is this a matter of just [changing the setting](https://github.com/blog/2224-change-the-base-branch-of-a-pull-request)?

Hussam-Eddin-Alhomsi commented:
I really like the idea behind the name of the plug-in that it handles mathematical expressions and the name describes it really well but I suggest we change the name of the plug-in from "**X-Pressive**" to "**X-pressive**" because capitalizing the the letter p in the first one makes it like **ExPressive** rather than "**Expressive**".

zonkmachine commented:
Tested the graph draw feature. Nice! You can't draw unless the smooth knob value is 0 as I understand it and the crossmark is inactive to reflect this. It works.

Umcaruje commented:
@zonkmachine I don't have time to test this, is this ready for a merge?

zonkmachine commented:
> @zonkmachine I don't have time to test this, is this ready for a merge?

I think so and @gnudles comment some 5 posts up reads:

> I think it is ready to be merge.

That's good enough for me.
@tresf merge?

Lukas-W commented:
Not good to merge. Due to the rebasing to `stable-1.2` this includes some commits from `master` that are not present in `stable-1.2`, such as de3b344edd4ebb7021b589869e8eaff69a8840d1 and 9f905bce3eee4e24a2338d9d8a104af6bdbec8ad.

Umcaruje commented:
>Not good to merge. Due to the rebasing to stable-1.2 this includes some commits from master that are not present in stable-1.2, such as de3b344 and 9f905bc.

Oh damn how did I miss this? @gnudles since your changes are mostly focused on a single folder, I would recommend backing up your changes, hard-resetting to `stable-1.2` and reapplying your changes on top of that. Note that you'll need to force-push

Umcaruje commented:
Or if someone else has something better in mind it'd be interesting to see how one can get out of this mess

Lukas-W commented:
I'd say we just merge it into `master`, as the it was intended originally. We'll never release 1.2 if we don't start taking "feature freeze" seriously. 

We can release 1.3 as quickly as possible after 1.2 is out (we could go into feature freeze pretty much immediately). We should move towards a much quicker release cycle anyway. Our last feature release 1.1.0 was in 2014!

Umcaruje commented:
>I'd say we just merge it into master, as the PR was intended originally. We'll never release 1.2 if we don't start taking "feature freeze" seriously.

Yeah I guess you're right.

zonkmachine commented:
So, bump this to 1.3?

zonkmachine commented:
@gnudles I looked through the commits and these are the ones that seem to be already in stable-1.2. Mind you, I'm not super savvy with git...

 `*` Marks merge commits


`Small fixes:	`								953d59ffd7c6126ed66dc0660f8ab9e137fc373b
`Added Buzzer Beater to demos (#3482)	`					32f07ac317b42dda42a1d00b0a3b1fa1feef0655
`Simpler SF2 Player file dialog. Copied from GIG Player. Opens correct…`		09c9ccb34dc6bb54954da9291ddfed3b02b4ebbd
`Fixed #3182. Always using master channel for preset previews. (#3503)` 		2df56829a31151f448acb48cdacbcc96c0a0e35e
`Dual Filter Plugin Redesign (#3484)`						2f51062d1e6b144a0ffe5229ba445f262d62ff3f
`EQ plugin now responds to wet / dry control`					a7154df004be92e1692b41f3967476eb0819b0a4
`Revert "'Off beat' beat notes to MelodyPattern" `				082dbed8e9eb7194c96e0f51e79b6e7d84032a03
 `* Merge pull request #3520 from curlymorphic/eqWetDry`				0190256801eab505967c43b44ddfc3f713451f31
` * Merge pull request #3522 from curlymorphic/delaySync`				77496188ae7c6e12f0b432e69db7b7325857de98
`Disable HiDPI on Windows`							d2e50df3ce2f8a56fd1750cc077eff3e76a6867d
` * Merge pull request #3531 from tresf/hidpi` 					6a5255e3795cf2079e3d1bebef5f32ef38876cff
`Make factory samples relative (#3510)` 						7e3ee14cf1a2d3e2975256c1fc2b8fb37d4c1c63
`LFO controller now has correct frequency with multiple connections. `		1848ed20d9802c3615138a3d67a883df8d7c34dc
`Fix automation processing in BB tracks (#3481)` 					75077f6200a5aee3a5821aae48a3b8466ed8714a
`Merge pull request #3535 from curlymorphic/lfo `					ff4ffd17f41ffd04ab179492868c40a55c591f35
`Implement 24 bit support for WAV export (#3021) `				4196a3041535010e7fa70c554cbd690a32a5ca91
`Simplification of code and structure in the area of the file export ` 		d5fc38ebd9c45000a745aa0b2dd8d5d25bf1a2fb
`Only show relevant widgets on the export dialog` 				4ac6aa909d5e804a8101e340fd5d1d88d2faa6ca
`Add the option to enable variable bit rates to the export dialog` 		823788919a1d3d411162eba425c6a6df1d03055e
`Add copyright info to the file OutputSettings.h` 				f06eac3f91bd77f51f5be1b560817ef8ab8fb84e
`Make the 24 bit export the default selection in the export dialog` 		4b3309a4e991d01542e980cd07237ae2106e01e3
`Rename entry for bit depth from "24 Bit Float" to "24 Bit Integer"` 		dfeb6bea3a415ffb9757ff4b7018cd7714bf0ee6
` * Merge pull request #3552 from michaelgregorius/3021-24-Bit-Export-For…` 	08201c202494c25ab8a151c98118dfa0359500e6
`Remove Limited Session` 								290556e43d200952e85a9f54607af9501621d484
`Provide support for fallback config values (#3551)` 				17b37e02a986c05b5a1af16f2852061b088a68eb
`don't change sampleTco length if tempo changed (#3543)` 				69d09f01865771c9604be002a3568d72ca9a90d2
`ReverbSC: Method to change samplerate (#3401) (#3403)` 				f390fd1723452f56740c4feb5ebcc2a6debd6da3
`Update Discord invite to link to #welcome (#3556)` 				b0ae31c2ffa96432e25ddbefe90b62270483cdae
`Remove glitching from the Flanger and Delay plugins (#3524)` 			b432707b9beba0bee1e37281bfa1c5e98e977172
`Flanger LFO rate syncronised (#3521)` 						300058df0430848697a9ef6f0175857822d4cdd9
`Replacement metronome samples (#3513)` 						c0b910ec578889065b8e46734bfdb36f14f74403
` * Merge branch 'stable-1.2' into xpressive_plugin`				887eeda6567a327cc73f458182b537f7f3d280cf

zonkmachine commented:
As the branch has already been switched back to master and this needs some more work, I bump this to the 1.3.0 milestone.

zonkmachine commented:
@tresf If I've understood this correctly the conflicts here will be easier to fix once https://github.com/LMMS/lmms/pull/3607 is merged, no?

tresf commented:
> If I've understood this correctly the conflicts here will be easier to fix once #3607 is merged, no?

Looks like that but I think it's just suffering from some branch switching.  Someone with some patience should be able to rebase against master again without losing work.  I say this because the conflicts above aren't actually edited in this PR.

zonkmachine commented:
> Someone with some patience...

OK. I've fixed it.
 * Branched master
 * Cherry-picked all commits on `gnudles:xpressive_plugin`...
   * that was from @gnudles except the first one, 953d59ffd7c6126ed66dc0660f8ab9e137fc373b, which is already in master.
   * that was from **ArashPartow** .
   * but omitted the merge commits.

Compiled, runs and still has the groove. :man_dancing: 

@gnudles @tresf Do you want me to push this to a separate branch or should I force push it over **gnudles/lmms** directly?

zonkmachine commented:
@tresf Fixed this but now there's a merge conflict over here: https://github.com/gnudles/lmms/pull/8

tresf commented:
> @tresf Fixed this but now there's a merge conflict over here: gnudles#8

Right, I guess we would expect that.  Any PRs against his PR would go into conflict.

> It looks like the easiest way is to reset this branch to before my fixup and then merge that one and cherry-pick it to my fixup branch?

The offending PR is on someone else's fork and was performed by @ArashPartow (the author of exprtk).  He's best to answer this question.  The unknown branch may be because this PR was handled elsewhere and he deleted his branch.  He's best to answer that question but at a glance, it appears you may be able to hand-edit them into this PR prior to merging.  The changes are minor.

zonkmachine commented:
```diff
-               typedef std::iterator_traits<typename generator::token_list_t::iterator>::difference_type diff_t;
+               typedef std::iterator_traits<generator::token_list_t::iterator>::difference_type diff_t;
```

tresf commented:
Oh, I see.   We if you've already incorporated them, the PR is moot.

zonkmachine commented:
```diff
-               typedef std::iterator_traits<typename generator::token_list_t::iterator>::difference_type diff_t;
+               typedef std::iterator_traits<generator::token_list_t::iterator>::difference_type diff_t;
```

Sorry, I pressed send there by mistake. The second (+) line above is indeed in [exprtk.hpp upstream](https://raw.githubusercontent.com/ArashPartow/exprtk/master/exprtk.hpp) but neither of the lines is anywhere in the versions I have. Either I've messed up or the left behind PR was never incorporated for a reason. Let's wait a couple of days for some feedback @gnudles and/or @ArashPartow.

zonkmachine commented:
> The second (+) line above is indeed in exprtk.hpp upstream but neither of the lines is anywhere in the versions I have.

_Edit: apart from the backed up original..._

gnudles commented:
@zonkmachine, you can always get the newest exprtk.hpp from it's repository. Thanks for the fixes BTW.
The exprtk.hpp in the plugin is the original file without any change, so it should be OK to take the newest.

zonkmachine commented:
Aight, upated! :wink:

zonkmachine commented:
> Not good to merge. Due to the rebasing to stable-1.2 this includes some commits from master that are not present in stable-1.2, such as de3b344 and 9f905bc.

I believe this is fixed now. Merge?

Lukas-W commented:
> Merge?

👍

zonkmachine commented:
@tresf This is a big one. Does anything get squashed or just merge it as it is?

tresf commented:
@zonkmachine if we wanted we could isolate the Graph additions and plugin cmake cleanup, but they're actually relatively small, I vote squash 'em all.

zonkmachine commented:
> I vote squash 'em all.

I'm fine with that but how about keeping the two main contributors apart and squashing the exprtk.hpp commits by Arash (amended by my update) separately so it gets recognized on the contributors list. I'm a bit sentimental that way. I tested it in a separate branch and the rebase comes out with no bugs and there is no diff against `gnudles:xpressive_plugin`.

It would look something like:

```
commit fa59472eed44f5e40d011bf3366874949720be89
Author: Orr Dvori <dvoreader@gmail.com>
Date:   Wed May 17 22:43:31 2017 +0300

    X-Pressive plugin - added graph drawing feature

commit acee8d3c6aefdcb7eaa047ce83881e3852ee2701
Author: ArashPartow <partow@gmail.com>
Date:   Sat Jan 14 05:21:23 2017 +1100

    Update ExprTk to tip of branch

commit 9b646c5c01266ef9229a6ff3f49c4ba32fcdb356
Author: Orr Dvori <dvoreader@gmail.com>
Date:   Sun Jan 8 14:31:01 2017 +0200

    X-Pressive Plugin
    
    Synthesizer plugin based on Bitinvader and 'The C++ Mathematical
    Expression Toolkit Library (ExprTk)'
    
    Some initial functions.
    
    available keys:
    f- note's frequency. available only in the output expressions
    t- time in seconds. in the Waves (W1,W2,W3) it's in the range [0,1)
    key- the note's keyboard key. available only in the output expressions.
    v- the note's velocity (divided by 255.0 so it is in the range [0,

...
```

tresf commented:
In my experience squash maintains the contributors.

zonkmachine commented:
Ah, OK! Then just a Squash and merge.

zonkmachine commented:
> In my experience squash maintains the contributors.

It doesn't [look like it](https://github.com/zonkmachine/lmms/graphs/contributors?from=2017-06-09&to=2017-06-26&type=a).

tresf commented:
What is the concern?  It gives him credit?

![image](https://user-images.githubusercontent.com/6345473/27553654-c9769cde-5a79-11e7-9ca0-01e4c940689d.png)

zonkmachine commented:
> What is the concern? It gives him credit?

Yes, gnudles but not Arash. It's not a big concern though, I'm more trying to get to grips with git and github.

ArashPartow commented:
@zonkmachine - Gnudles did all the heavy lifting, I'm if I don't get through.  :D

gnudles commented:
Maybe we should credit Arash in the plugin itself in the help window.
BTW, shouldn't it be merged to stable1.2?

zonkmachine commented:
> BTW, shouldn't it be merged to stable1.2?

Agree. In the works here: https://github.com/LMMS/lmms/pull/3845

Unesty commented:
How to add external sample to W?

gnudles commented:
you have the load button

Unesty commented:
I tried to load sample and it became distorted. Is that how supposed to be?

gnudles commented:
the wave is loaded to a buffer with 4096 samples or so (4096 floats or shorts, can't recall). 4096 is a huge buffer compared to other plugins in lmms. but if you load a long sample it will become distorted (resampled). it is not a bug, it's a feature :)
